# Hands-on Exercises in Machine Learning Applications in Digital Cultural Heritage
# Part II -- Computer Vision

### Tentative Schedule:
* 1 - Vectors and Matrices

* 2 - ANN

* 3 - Convolutions

* 4 - Convolutions 2

* 5 - Transformers intro?

* 6 - Classification

* 7 - Detection (YOLO?)

* 8 - Segmentation (SAM?)

* 9 - Autoencoders

* 10 - Autoencoders 2?

* 11 - Review

* 12 - Bias/Black Box?

* 13 - Cool stuff

* 14 - Student Presentations

### The Goal

ML Computer Vision techniques are increasingly finding their way into Cultural Heritage papers. To keep up with the research you need to have a basic understanding of how these work.

For example, by the end of the course it would be good to understand photos such as the below one (from a paper on image restoration).

<div>
<img src="./Reconstruction.png" width="1000"/>
</div>
Source: https://arxiv.org/abs/2108.10257

# ML Week 11*: Vectors and Matrices

##### So far we have only dealt with tabular and point cloud data. What about Image and Video data?

##### To deal with these, we need to learn the basics of Matrices and Vectors (a domain known as Linear Algebra).

##### Be brave! This is nothing scary. The only operations you need to know are addition and multiplication.

\* Note about naming convention: Continuation from last semester's ML course, hence not week 1. Except there was no "week 10" course. But starting at 11 (1 mod 10) just feels more natural.

## We are now leaving the domain of Pandas DataFrames behind

## It's Numpy from here on out...

In [2]:
import numpy as np

## Let's start with Vectors

The following code initializes a basic "row vector", with the form:
$\left[1,2,3,4\right]$

In [3]:
v = np.array([[1,2,3,4]])

Vectors can also be "column vectors". These have the form: $\begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}$

In [4]:
w = np.array(
    [
    [1],
    [2],
    [3],
    [4]
    ]
)

What is the difference between a row vector and a column vector? When combining them with other vectors or matrices (usually using multiplication) the order in which we carry out the multiplications can change. We will see this later.

Note: It's also possible to define a row vector with only one pair of squre bracket. For many purposes, this is fine. But be careful: the ambiguity can cause issues sometimes.

In [5]:
v2 = np.array([1,2,3,4])

In [6]:
# We can see the differences in the shapes
print(v.shape)
print(w.shape)
print(v2.shape)

(1, 4)
(4, 1)
(4,)


The transpose method (simply "T" in numpy) flips a vector from a row vector to a column vector:

In [7]:
v.T

array([[1],
       [2],
       [3],
       [4]])

In [8]:
w.T

array([[1, 2, 3, 4]])

## Now let's make a Matrix!

The following code will initialize a matrix of the form $\begin{bmatrix} 1 & 2 & 3 & 4 \\ 5 & 6 & 7 & 8 \\ 9 & 10 & 11 & 12 \\ 13 & 14 & 15 & 16 \end{bmatrix}$

In [9]:
M = np.array(
    [
    [1,2,3,4],
    [5,6,7,8],
    [9,10,11,12],
    [13,14,15,16]
    ]
)

In [10]:
M

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [11]:
# You can also transpose a matrix (this is equivalent to flipping it diagonally)
M.T

array([[ 1,  5,  9, 13],
       [ 2,  6, 10, 14],
       [ 3,  7, 11, 15],
       [ 4,  8, 12, 16]])

In [12]:
# Vectors and Matrices can be initialized with random entries:
np.random.rand(4,4)

array([[0.89165701, 0.52945198, 0.41828402, 0.64796091],
       [0.91951986, 0.08268713, 0.43090811, 0.95206904],
       [0.80440653, 0.66016571, 0.15597671, 0.00361193],
       [0.25106557, 0.00574373, 0.68785873, 0.43158227]])

## Now let's explore addition and multiplication

Things are quite simple if both the vectors are row vectors, or if both the vectors are column vectors:

In [13]:
v + v

array([[2, 4, 6, 8]])

In [14]:
v * v

array([[ 1,  4,  9, 16]])

In [15]:
w + w

array([[2],
       [4],
       [6],
       [8]])

In [16]:
w * w

array([[ 1],
       [ 4],
       [ 9],
       [16]])

But things get more complicated if we have a row and a column vector (or a vector and a matrix, or two matrices).

In such cases, the manner in which we should add or multiply the individual numbers is somewhat ambiguous.

In general, there are 3 different ways we can add/multiply in these situations. They are called:

* Broadcasting (addition, subtraction, multiplication, or division)
* Dot product (multiplication)
* Cross product (multiplication)

Here we will see broadcasting and the dot product. These form the foundation of how neural networks are built!

## Broadcasting

Let's first look at vector-vector broadcasting.

$[1,2,3,4] + \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}$

or

$\begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix} + [1,2,3,4]$

In [17]:
v + w

array([[2, 3, 4, 5],
       [3, 4, 5, 6],
       [4, 5, 6, 7],
       [5, 6, 7, 8]])

In [18]:
w + v

array([[2, 3, 4, 5],
       [3, 4, 5, 6],
       [4, 5, 6, 7],
       [5, 6, 7, 8]])

$[1,2,3,4] * \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}$

or

$\begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix} * [1,2,3,4]$

In [19]:
v * w

array([[ 1,  2,  3,  4],
       [ 2,  4,  6,  8],
       [ 3,  6,  9, 12],
       [ 4,  8, 12, 16]])

In [20]:
w * v

array([[ 1,  2,  3,  4],
       [ 2,  4,  6,  8],
       [ 3,  6,  9, 12],
       [ 4,  8, 12, 16]])

Now let's see vector matrix broadcasting:

In [21]:
v*M

array([[ 1,  4,  9, 16],
       [ 5, 12, 21, 32],
       [ 9, 20, 33, 48],
       [13, 28, 45, 64]])

In [22]:
M*v

array([[ 1,  4,  9, 16],
       [ 5, 12, 21, 32],
       [ 9, 20, 33, 48],
       [13, 28, 45, 64]])

In [23]:
w*M

array([[ 1,  2,  3,  4],
       [10, 12, 14, 16],
       [27, 30, 33, 36],
       [52, 56, 60, 64]])

In [24]:
M*w

array([[ 1,  2,  3,  4],
       [10, 12, 14, 16],
       [27, 30, 33, 36],
       [52, 56, 60, 64]])

## Dot Product
The dot product does something similar to broadcasting, but for each **row** you multiply with every **column**, and then all the muliplied elements are added together.

In [ ]:
# When you try to do a dot product between a column vector and a row vector it turns out to be the same as the normal broadcasting (w*v)
w @ v

array([[ 1,  2,  3,  4],
       [ 2,  4,  6,  8],
       [ 3,  6,  9, 12],
       [ 4,  8, 12, 16]])

In [ ]:
# But if you do a dot product between a row vector and a column vector then get something quite interesting:
v @ w

array([[30]])

This is because:
$$[1,2,3,4] \cdot \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix} = 1*1 + 2*2 + 3*3 + 4*4 = 1 + 4 + 9 + 16 = 30$$

When using the dot product to multiply vectors and matricies, it will only work if the dimensions match (the length of the first "thing" must match the height of the second "thing"):

In [ ]:
# v @ M will work because the length of v (4) matches the height of M (4)
print(str(v.shape) + " -> " + str(M.shape))

(1, 4) -> (4, 4)


In [25]:
v @ M

array([[ 90, 100, 110, 120]])

$$ [1,2,3,4] \cdot \begin{bmatrix} 1 & 2 & 3 & 4 \\ 5 & 6 & 7 & 8 \\ 9 & 10 & 11 & 12 \\ 13 & 14 & 15 & 16 \end{bmatrix} = \left[ 1*1 + 2*5 + 3*9 + 4*13, 1*2 + 2*6 + 3*10 + 4*14, 1*3 + 2*7 + 3*11 + 4*15, 1*4 + 2*8 + 3*12 + 4*16 \right]$$
$$ = [90,100,110,120]$$

In [45]:
# w @ M will NOT work because the length of w (1) matches the height of M (4)
print(str(w.shape) + " -> " + str(M.shape))

(4, 1) -> (4, 4)


In [34]:
w @ M

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 4 is different from 1)

In [ ]:
# But M @ w will work because length of M (4) matches height of w (4)
M @ w

array([[ 30],
       [ 70],
       [110],
       [150]])

$$\begin{bmatrix} 1 & 2 & 3 & 4 \\ 5 & 6 & 7 & 8 \\ 9 & 10 & 11 & 12 \\ 13 & 14 & 15 & 16 \end{bmatrix} \cdot \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix} = \begin{bmatrix} 1*1 + 2*2 + 3*3 + 4*4 \\ 5*1 + 6*2 + 7*3 + 8*4 \\ 9*1 + 10*2 + 11*3 + 12*4 \\ 13*1 + 14*2 + 15*3 + 16*4 \end{bmatrix} = \begin{bmatrix} 30 \\ 70 \\ 110 \\ 150 \end{bmatrix}$$

# Exercises

### Exercise 1
We sometimes want to take a small section of a matrix to study it further (this will be quite important in week 3).

Initialize a matrix with random entries, and shape (5,5).

Then extract a smaller (2,2) matrix from the top left corner.

Recall you can use the square brackets [ ] to index matrices.

### Exercise 2

Create a matrix containing only zeros with shape (4,4). Hint: look into the np.zeros method.

Set all of the diagonal entries to 1 (the diagonal is defined as going from the top left of the matrix to the bottom right).

This special matrix is called the "Identity Matrix", often defined as "I". It has the property that it can multiply any matrix and the result will always be the unchanged matrix.

$$ I\cdot M = M\cdot I = M $$

Randomly define another 4x4 matrix of integers (hint: look up np.random.randints()). Then multiply them together to check that this property is indeed true.

Take a moment to understand why this is the case.

### Exercise 3

A vector can represent the probability of an image belonging to certain classes.

For example, a vector $[0.1,0.6,0.1,0.2]$ would be interpreted as the image having a 10% chance of belong to class 1, a 60% chance of belonging to class 2, a 10% chance of belonging to class 3, and a 20% chance of belonging to class 4. In such a case, to make a decision, we are only interested in the class with maximum probability.

Initialize a vector with length 10 using np.random.rand()

Then find a way to manipulate it such that the maximum entry is set to 1 and all others are set to 0 (you may need to look online).

We will see this more in the next class.